### ISSUE TITLE
[Bug] alignment between timescale and timeline breaks in viewmodes week and month

### ISSUE THREAD

**Issue by Klaus-Valse** [frappe/gantt#417]
# Description of the error
The timeline (the area that holds all tasks) is not aligned with the time scale (the area that holds days/weeks/months/etc. and the coresponding numeric values) in the week and month viewmode. The day viewmode works perfectly fine and I could not break it. I have not extensively investigated the behaviour on the year mode, since it is very coarse for my usecase, so the error also might occur there.

<img width="295" alt="bug" src="https://github.com/user-attachments/assets/362b19a7-54c2-45dd-bb18-3ba036a44dd0">

# What I discovered so far
If the task are relatively narrow in terms of their dates, which means that they are all starting and ending in the same month, the error is not really visible. However the further the tasks are apart, or the larger the range of the entire chart is, the more severe the error gets _up to a certain point_ and now it gets weird: I changed the last task's start and end dates to 2023 instead of 2024 to influence the shift in the week viewmode and I could. Then I changed it to 2020 instead of 2024 to see if the shift increases and suddenly no shift at all in the week viewmode. In both cases there was a shift in the month view mode though. Same if I try 2019. 2018 and there is the shift again.
When changing from day to week viewmode, the error does not occur, the blue today-line is perfectly in the middle of the blue bubble. To make it break you need to scroll all the way to the **left**. I tried scrolling to the right to provoke this, when I discovered something else: the timeline and the timescale are shifted alltogether _if_ the shift occurs.  
This is what it looks like when changing the last task to 2021:
<img width="803" alt="2024-08-07 18_48_39-2021 bad" src="https://github.com/user-attachments/assets/a5969b68-75ed-4bdc-9863-9448f2ead542">  

And this is what it looks like when changing the last task to 2020:
<img width="812" alt="2024-08-07 18_48_39-2020 alright" src="https://github.com/user-attachments/assets/0e93cc30-1932-439f-a7ea-62caa4a1b6f4">  

When I scrolled all the way to the left I noticed if I slowly scroll back, the timeline moves accordingly but the timescale remains still. only after some time scrolling back to the right, the timescale starts moving too.

# Code for reproducing the error
```
<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>Simple Gantt</title>
    <style>
      body {
        font-family: sans-serif;
        background: #ccc;
      }
      .container {
        width: 80%;
        margin: 0 auto;
      }
      .heading {
        text-align: center;
      }
      .gantt-target.dark {
        background-color: #252525;
      }
    </style>
    
    <script src="frappe-gantt.min.js"></script>
    <link rel="stylesheet" href="frappe-gantt.css">
  </head>
  <body>
    <div class="container">
      <h2 class="heading" id="dateHeading">Today: YYYY-MM-DD</h2>
      <div class="gantt-target"></div>
    </div>
    <script>
      const tasks = [
          {
              "dependencies": [],
              "end": "2024-11-29",
              "id": "ID-41687",
              "name": "task 1 (2024-03-14 - 2024-11-29)",
              "progress": 100,
              "start": "2024-03-14"
          },
          {
              "dependencies": ["ID-41687"],
              "end": "2024-08-14",
              "id": "ID-44368",
              "name": "task 2 (2024-08-07 - 2024-08-14)",
              "progress": 100,
              "start": "2024-08-07"
          },
          {
              "dependencies": ["ID-41687"],
              "end": "2024-08-14",
              "id": "ID-41606",
              "name": "task 3 (2024-08-05 - 2024-08-14)",
              "progress": 100,
              "start": "2024-08-05"
          },
          {
              "dependencies": ["ID-40403"],
              "end": "2024-08-29",
              "id": "ID-41323",
              "name": "task 4 (2024-08-25 - 2024-08-29)",
              "progress": 100,
              "start": "2024-08-25"
          },
          {
              "dependencies": ["ID-40403"],
              "end": "2024-08-15",
              "id": "ID-39762",
              "name": "task 5 (2024-08-14 - 2024-08-15)",
              "progress": 100,
              "start": "2024-08-14"
          },
          {
              "dependencies": ["ID-36739"],
              "end": "2024-06-14",
              "id": "ID-38216",
              "name": "task 6 (used for spreading the range)",
              "progress": 100,
              "start": "2024-03-25"
          }
      ];

      document.addEventListener("DOMContentLoaded", function () {
          const today = new Date().toISOString().slice(0, 10);
          const heading = document.getElementById("dateHeading");
          heading.innerText = `Today: ${today}`;
          let gantt_chart = new Gantt(".gantt-target", tasks, {
          on_click: (task) => {
              console.log("Click", task);
          },
          on_double_click: (task) => {
              console.log("Double Click", task);
          },
          on_date_change: (task, start, end) => {
              console.log("Date change", task, start, end);
          },
          on_progress_change: (task, progress) => {
              console.log("Progress Change", task, progress);
          },
          on_view_change: (mode) => {
              console.log("View Change", mode);
          },
          on_hover: (task, x, y) => {
              console.log("Hover", x, y);
          },
          view_mode: "Day",
          view_mode_padding: { DAY: "3d" },
          popup: false,
          scroll_to: 'today',
          view_mode_select: true,
          today_button: true,
          readonly: false,
          lines: 'vertical'
          });
      });
    </script>
  </body>
</html>
```

# Environment
Tested on Windows 10 with Chrome, Firefox and Edge - all show the same behaviour here.
Tested on MacOS Monterey with Chrome and Brave - same behaviour however breaking it with scrolling to the left was harder. In both browsers it did not visibly occur until I spread the chart by moving one task to 2023 or earlier.

This is a major issue that I discovered pretty late in development of my own gantt chart based on Frappé Gantt, so I did some investigation and will try my best to fix it on my own (because this bug it is very severe and urgent for me). Any help is much appreciated.

**Comment by Klaus-Valse** [comment]
### Update

I checked out the version from 6 months ago, before the last big update (https://github.com/frappe/gantt/tree/be1c8752221d03938a8155833ca02d29b74bd15c)

and there the problem does not occur. The new dynamic month display seems suspicious, because in the old version everything is already rendered and responding directly to the scrollbar movement, whereas the new version does in some cases (week or month view, when you scroll all the way to the left - see details above).

**Comment by safwansamsudeen** [comment]
Thank you for your detailed report! We'll fix this soon.

**Comment by Klaus-Valse** [comment]
Hey!

I fixed the problem temporarily by rolling back to the version before the April 2024 update.
Everything worked like a charm again until the user of my adaptation loaded data spanning over 3 or more years. Then the month view (which of course was the most needed one for our usecase) caused later bars to be shifted. I think I fixed the error now and I checked, it still seems to be in the current code base, so here is what I did:

* in Frappé's bar.js, I changed compute_x():  
  ```
  compute_x() {
        const { step, column_width } = this.gantt.options;
        const task_start = this.task._start;
        const gantt_start = this.gantt.gantt_start;

        const diff = date_utils.diff(task_start, gantt_start, 'hour');
        let x = (diff / step) * column_width;

        /* Since the column width is based on 30,
        we count the month-difference, multiply it by 30 for a "pseudo-month"
        and then add the days in the month, making sure the number does not exceed 29
        so it is within the column */
        if (this.gantt.view_is('Month')) {
            const diffDaysBasedOn30DayMonths = date_utils.diff(task_start, gantt_start, 'month') * 30;
            const dayInMonth = Math.min(29, date_utils.format(task_start, "DD"));
            const diff = diffDaysBasedOn30DayMonths + dayInMonth;
            
            x = (diff * column_width) / 30;
        }
        
        return x;
    }
  ``` 
  
* Furthermore, date_utils.diff() did not always yield correct results, so I changed it to:  
  ```
  diff(date_a, date_b, scale = DAY) {
    let milliseconds, seconds, hours, minutes, days, months, years;

    milliseconds = date_a - date_b;
    seconds = milliseconds / 1000;
    minutes = seconds / 60;
    hours = minutes / 60;
    days = hours / 24;

    // Calculate months across years
    const yearDiff = date_a.getFullYear() - date_b.getFullYear();
    const monthDiff = date_a.getMonth() - date_b.getMonth();

    /* If monthDiff is negative, date_b is in an earlier month than
    date_a and thus subtracted from the year difference in months */
    months = yearDiff * 12 + monthDiff;

    /* If date_a's (e.g. march 1st) day of the month is smaller than date_b (e.g. february 28th),
    adjust the month difference */
    if (date_a.getDate() < date_b.getDate()) {
        months--;
    }

    // Calculate years based on actual months
    years = months / 12;

    if (!scale.endsWith("s")) {
      scale += "s";
    }

    return Math.floor(
      {
        milliseconds,
        seconds,
        minutes,
        hours,
        days,
        months,
        years,
      }[scale]
    );
  }
  ```

* In Frappé's gantt class, the make_grid_ticks() function did not calculate the position of the month labels correctly:
  ```
  make_grid_ticks() {
        .
        .
        .
            if (this.view_is(VIEW_MODE.MONTH)) {
                tick_x +=
                    (date_utils.get_days_in_month(date) *
                        this.options.column_width) /
                    30;
            } else {
                tick_x += this.options.column_width;
            }
        }
    }
  ```  
  This method was overridden and this, what appears to be unnecessarily complex way to caluclate the position has been simpliefied to:
  ```
  make_grid_ticks() {
        .
        .
        .
        tick_x += this.options.column_width;
    }
  ```  
  
  So in conclusion I think it's best to work with "pseudo-months" of 30 days consistently over the entire codebase.
Thanks for your efforts and help! Without the thorough documentation and the clear design this would not be as easy to work with.

**Comment by Klaus-Valse** [comment]
@safwansamsudeen, please check my comment above, it might save some time investigating. Cheers

**Comment by safwansamsudeen** [comment]
@Klaus-Valse thanks for this catch! It turned out to be a rather nifty bug, and the solution was fairly simple. Should be fixed now!
<img width="1150" alt="image" src="https://github.com/user-attachments/assets/1d44f6ca-f236-437b-8011-61a7a2fce4e1">
